### PR TITLE
Scylla server rust 2: Recievers and handlers

### PR DESCRIPTION
**Pull request by jr1221**
Split stuff with different terminology bc the proxy stuff confused me.

Recievers: things that recieve data and parse it into ClientData type: mqtt_reciever, mock (tbd)

Handler: things that take the recieved data as ClientData and send it places: socket_handler, db_handler

TODO to make same functionality as JS:

- [x] mock reciever
- [x] latency measure
- [x] make the disconnected thing work in client??
- [x] add verbosity to connection to siren

### CODE REVIEW

File: scylla-server-rust/src/services/data_service.rs [hunk 1]
```diff
@@ -1,28 +1,36 @@
...
-/// * `run_id` - The run id to filter the data by
+/// * `run_id` - The run id to filter the data
+/// * `fetch_run` whether to fetch the run assocaited with this data
+/// * `fetch_data_type` whether to fetch the data type associated with this data
 /// returns: A result containing the data or the QueryError propogated by the db
 pub async fn get_data(
     db: &Database,
     data_type_name: String,
     run_id: i32,
+    fetch_run: bool,
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/services/data_service.rs]: why are we adding these these seem unnecessary

File: docker-compose-dev.yml [hunk 2]
```diff
@@ -1,79 +0,0 @@
...
-      - 1883:1883
-      - 9001:9001 # why?
-    expose:
-      - 1883
-    volumes:
-      - ./siren-base/mosquitto/mosquitto.conf:/mosquitto/config/mosquitto.conf
-    cpu_shares: 2048
-    cpuset: 0-3
-    mem_limit: 2gb
-    oom_kill_disable: true
```
Comment (RChandler234) [inline docker-compose-dev.yml:76]: does the other compose work for router and local now?

File: .github/workflows/scylla-rust-build.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+
+
+# Defines two custom environment variables for the workflow. These are used for the Container registry domain, and a name for the Docker image that this workflow builds.
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}
+
+# There is a single job in this workflow. It's configured to run on the latest available version of Ubuntu.
+jobs:
```
Comment (RChandler234) [inline .github/workflows/scylla-rust-build.yml]: why two separate build actions now?

File: scylla-server-rust/README.md [hunk 4]
```diff
@@ -24,7 +24,16 @@ SOURCE_DATABASE_URL=postgresql://postgres:password@127.0.0.1:5432/timescaledb ca
...
 docker volume rm argos_db-data
-docker compose run -Pd odyssey-timescale
+docker compose up odyssey-timescale
 cargo prisma migrate deploy
 SOURCE_DATABASE_URL=postgresql://postgres:password@127.0.0.1:5432/timescaledb cargo test -- --test-threads=1
-```
\ No newline at end of file
+```
+
+### Deploy this app
```
Comment (RChandler234) [inline scylla-server-rust/README.md:70]: I've been keeping deploy info in the main README, and then setup info in these sub ones. I would at least make sure this info is reflected in the main README

File: scylla-server-rust/src/main.rs [hunk 5]
```diff
@@ -7,13 +7,83 @@ use scylla_server_rust::{
...
+        tx,
+        std::env::var("PROD_SIREN_HOST_URL").unwrap_or("localhost:1883".to_string()),
+        db.clone(),
+        io,
+    )
+    .await;
+    tokio::spawn(recv.recieve_mqtt(eloop));
 
     let app = Router::new()
         // get all data with the name dataTypeName and runID as specified
```
Comment (RChandler234) [inline scylla-server-rust/src/main.rs]: sorry I didn't follow up on my comment on these last time. These comments should be separators that group routes that correspond to the same controller methods together. So here would be something like: /********** Data Routes ********/. These comments you have here now is info someone should be able to get from reading the route and controller name that's being called

File: scylla-server-rust/src/reciever/mqtt_reciever.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
+use core::fmt;
```
Comment (RChandler234) [inline scylla-server-rust/src/reciever/mqtt_reciever.rs]: you misspelled receiver in the folder/ file names

File: scylla-server-rust/src/reciever/mock_reciever.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                    .gen_range((self.min)..(self.max))
+                    .to_string(),
+            );
+        }
+
+        val_vec
+    }
+}
+
+const BASE_MOCK_DATA: [MockData; 17] = [
```
Comment (RChandler234) [inline scylla-server-rust/src/reciever/mock_reciever.rs]: it'd be nice if we could house the actual data that's sent in other files, so we can separate the logic/ mock data, which would hopefully make it a little easier to add more mock data/ organize it a lil better. Not sure how annoying that is in rust tho

File: scylla-server-rust/src/reciever/mqtt_reciever.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            unit: data.unit,
+            values: data.value,
+            timestamp: time_clean,
+            node: node.to_string(),
+        })
+    }
+
+    /// Send a message to the channel, printing and IGNORING any error that may occur
+    /// * `client_data` - The cliet data to send over the broadcast
+    async fn send_msg(&self, client_data: ClientData) {
```
Comment (RChandler234) [inline scylla-server-rust/src/reciever/mqtt_reciever.rs]: feels a little weird calling this just a receiver if it's in charge of receiving mqtt messages and sending data to client (if I'm reading this right)

File: scylla-server-rust/src/reciever/db_handler.rs [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    /// The list of data types seen by this instance, used for when to upsert
+    datatype_list: Vec<String>,
+    /// The broadcast channel which provides serial datapoints for processing
+    reciever: Receiver<ClientData>,
+    /// The database
+    db: Database,
+    /// An internal state of an in progress location packet
+    loc_lock: LocLock,
+    /// Whether the location has been modified this loop
+    is_loc: bool,
```
Comment (RChandler234) [inline scylla-server-rust/src/reciever/db_handler.rs]: can we call these location_lock / is_location, loclock feels a lil confusing for no reason

**Review by RChandler234 — CHANGES_REQUESTED** [review]

File: .github/workflows/scylla-rust-build.yml (hunk 3 above)
Comment (jr1221) [reply .github/workflows/scylla-rust-build.yml]: So we can save time on running one versus the other, so a person who needs a client rebuild doesnt have to wait 1.5 hours for a server rebuild.  I thought I asked u before changing it but perhaps I said something different lol.

File: docker-compose-dev.yml (hunk 2 above)
Comment (jr1221) [reply docker-compose-dev.yml:76]: right now the composes are kinda messed up.  basically there are three use cases:
- Dev (client points to localhost)
- Person running prod locally or on router w/ siren (the usual config)
- Person running prod on car (no client, siren, and scylla points to localhost siren).
I think I will need a compose for each one, or we have the user edit the composes.  Ill clean that up.

File: scylla-server-rust/src/reciever/mqtt_reciever.rs (hunk 8 above)
Comment (jr1221) [reply scylla-server-rust/src/reciever/mqtt_reciever.rs]: I refactored to "processor" lmk if that sounds better.

File: scylla-server-rust/src/main.rs (hunk 5 above)
Comment (jr1221) [reply scylla-server-rust/src/main.rs]: I still dont really get what you mean tbh sorry.  I made a new commit as follows:
```rust
// DATA ROUTES
        .route(
            "/data/:dataTypeName/:runId",
            get(controllers::data_controller::get_data),
        )
        // DATA TYPE ROUTES
        .route("/datatypes", get(data_type_controller::get_all_data_types))
        // DRIVERS
        .route("/drivers", get(driver_controller::get_all_drivers))
        // LOCATIONS
        .route("/locations", get(location_controller::get_all_locations))
        // NODES
        .route("/nodes", get(node_controller::get_all_nodes))
        // RUNS
        .route("/runs", get(run_controller::get_all_runs))
        .route("/runs/:id", get(run_controller::get_run_by_id))
        // SYSTEMS
        .route("/systems", get(system_controller::get_all_systems))
```

File: .github/workflows/scylla-rust-build.yml (hunk 3 above)
Comment (Peyton-McKee) [reply .github/workflows/scylla-rust-build.yml]: is there documentation on how to manually build and upload potentially?

File: docker-compose.yml [hunk 10]
```diff
@@ -17,34 +17,34 @@ services:
     mem_limit: 3gb
     stop_grace_period: 1m
 
-  scylla-server:
-    container_name: scylla-server
-    restart: unless-stopped
-    image: ghcr.io/northeastern-electric-racing/argos:Develop-scylla-server
+  scylla-server-rust:
+    container_name: scylla-server-rust
+    image: ghcr.io/northeastern-electric-racing/argos:rust-rewrite-socket-scylla-server-rust
```
Comment (Peyton-McKee) [inline docker-compose.yml]: does this image name have to be changed on every dev branch?

File: docker-compose.yml [hunk 11]
```diff
@@ -17,34 +17,34 @@ services:
...
+    stop_signal: SIGINT
+    init: true
+    entrypoint: ["./docker_run.sh"]
 
 
   client:
     container_name: client
     restart: unless-stopped
-    image: ghcr.io/northeastern-electric-racing/argos:Develop-angular-client
+    image: ghcr.io/northeastern-electric-racing/argos:develop-angular-client
```
Comment (Peyton-McKee) [inline docker-compose.yml]: what is the significance of the :develop-angular-client. is that a branch name?

File: scylla-server-rust/src/main.rs [hunk 12]
```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
...
+    // TODO tune buffer size
+    let (tx_proc, rx_proc) = mpsc::channel::<Vec<ClientData>>(1000);
+
+    // the below two threads need to cancel cleanly to ensure all queued messages are sent.  therefore they are part of the a task tracker group.
+    // create a task tracker and cancellation token
+    let task_tracker = TaskTracker::new();
+    let token = CancellationToken::new();
+    // spawn the database handler
+    task_tracker.spawn(
+        db_handler::DbHandler::new(rx, Arc::clone(&db)).handling_loop(tx_proc, token.clone()),
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/main.rs]: wouldnt cloning the token make a new token? therefore these are on two different tokens?

File: scylla-server-rust/src/main.rs [hunk 13]
```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
...
+
+    // create the socket stuff
+    let (socket_layer, io) = SocketIo::new_layer();
+    io.ns("/", |s: SocketRef| {
+        s.on_disconnect(|_: SocketRef| debug!("Socket: Client disconnected from socket"))
+    });
+
+    // channel to pass the mqtt data
+    // TODO tune buffer size
+    let (tx, rx) = mpsc::channel::<ClientData>(10000);
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/main.rs]: tx, rx mean nothing to me, should be mqttReceive and mqttSend

File: scylla-server-rust/src/main.rs [hunk 14]
```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
...
+
+    // channel to pass the processed data to the db thread
+    // TODO tune buffer size
+    let (tx_proc, rx_proc) = mpsc::channel::<Vec<ClientData>>(1000);
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/main.rs]: same thing here dbReceive, dbSend

File: scylla-server-rust/src/main.rs [hunk 15]
```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
...
+            tx,
+            std::env::var("PROD_SIREN_HOST_URL").unwrap_or("localhost:1883".to_string()),
+            db.clone(),
+            io,
+        )
+        .await;
+        tokio::spawn(recv.process_mqtt(eloop));
+    }
 
     let app = Router::new()
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/main.rs:130]: this still isnt making new routers.

File: scylla-server-rust/src/main.rs [hunk 16]
```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
...
+    info!("Use Ctrl+C or SIGINT to exit cleanly!");
 
-    let listener = tokio::net::TcpListener::bind("0.0.0.0:8000").await.unwrap();
-    axum::serve(listener, app).await.unwrap();
+    // listen for ctrl_c, then cancel, close, and await for all tasks in the tracker.  Other tasks cancel vai the default tokio system
+    signal::ctrl_c()
+        .await
+        .expect("Could not read cancellation trigger (ctr+c)");
+    info!("Received exit signal, shutting down!");
+    token.cancel();
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/main.rs:193]: do you have to also cancel the axum_token

File: scylla-server-rust/src/main.rs [hunk 17]
```diff
@@ -7,44 +7,153 @@ use scylla_server_rust::{
...
+    task_tracker.wait().await;
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/main.rs:194]: this is a beautiful file

File: scylla-server-rust/src/processors/mqtt_processor.rs [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+                Cow::Owned((
+                    "ts".to_string(),
+                    chrono::offset::Utc::now().timestamp_millis().to_string(),
+                ))
+            })
+            .into_owned();
+
+        // parse time, if time isnt present use sys time (see above)
+        let Ok(time_clean) = unix_time.1.parse::<i64>() else {
+            return Err(format!("Invalid timestamp: {}", unix_time.1));
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/processors/mqtt_processor.rs]: i dont think we should error, we should just use the router time

File: scylla-server-rust/src/processors/mqtt_processor.rs [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        };
+        // ts check for bad sources of time which may return 1970
+        if time_clean < 963014966000 {
+            return Err(format!("Timestamp before year 2000: {}", unix_time.1));
```
Comment (Peyton-McKee) [inline scylla-server-rust/src/processors/mqtt_processor.rs]: same here

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
This looks too lit. The public data stuff with the selects looks much much cleaner imo

File: docker-compose.yml (hunk 10 above)
Comment (jr1221) [reply docker-compose.yml]: Yup.  I'll see if there's a way around that but kinda sucks.  Basically user should always check the URLd to see if it's the branch they want.

File: docker-compose.yml (hunk 11 above)
Comment (jr1221) [reply docker-compose.yml]: Yup.  The develop is branch name and angular-client is item type.

File: scylla-server-rust/src/main.rs (hunk 12 above)
Comment (jr1221) [reply scylla-server-rust/src/main.rs]: You have to clone it to use it in the thread.  The clone keeps the link to the cancelation, tbh idk how it works but docs do say to clone.

File: scylla-server-rust/src/main.rs (hunk 15 above)
Comment (jr1221) [reply scylla-server-rust/src/main.rs:130]: I still don't understand the benefits to make a new router per 😢 .  Basically a new router in axum would not inherit state and CORS, meaning I need a seperate database reference and CORS layer for each router, resulting in a huge amount of ugliness and whatnot.  The closest I can get to what you are describing is https://github.com/tokio-rs/axum/issues/1326#issue-1350670719 which explains to create and clone an app state then merge them together.

File: scylla-server-rust/src/main.rs (hunk 16 above)
Comment (jr1221) [reply scylla-server-rust/src/main.rs:193]: Nah as it's cloned from the main one and becomes owned by the thread.  It's moved into the thread and the main token is used to cancel.
https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html#:~:text=Struct%20tokio_util%3A%3Async%3A%3ACancellationToken&text=A%20token%20which%20can%20be,resolved%20when%20cancellation%20is%20requested.

File: scylla-server-rust/src/processors/mqtt_processor.rs (hunk 19 above)
Comment (jr1221) [reply scylla-server-rust/src/processors/mqtt_processor.rs]: What if router time is 1970 also?  That's what this logic is for.

File: scylla-server-rust/src/main.rs (hunk 5 above)
Comment (RChandler234) [reply scylla-server-rust/src/main.rs]: yeah this is perfect, just to break up the routes visually. Sorry for the confusion, it was a nitpicky comment

File: .github/workflows/scylla-rust-build.yml (hunk 3 above)
Comment (jr1221) [reply .github/workflows/scylla-rust-build.yml]: I'll look into it.  It's just `docker compose build && push` but the github permissions idk about.

File: scylla-server-rust/src/reciever/mqtt_reciever.rs (hunk 8 above)
Comment (RChandler234) [reply scylla-server-rust/src/reciever/mqtt_reciever.rs]: Maybe manager? Like mqtt_manager, db_manager, mock_mqtt_manager (or mqtt could be socket_manager, mock_socket_manager up to you). Processor just feels like such a generic name, and manager describes what the service is doing (managing the connection with another process).

Comment (RChandler234) [reply scylla-server-rust/src/reciever/mqtt_reciever.rs]: Ultimately do whatever seems most intuitive to you, this is kind of a pedantic comment anyways

File: scylla-server-rust/src/main.rs (hunk 15 above)
Comment (RChandler234) [reply scylla-server-rust/src/main.rs:130]: ^^ I agree with Jack, I don't see the benefit when Scylla isn't going to have nearly as many routes as something like FinishLine (at least not soon). It feels like the technical lift would outweigh the negligible organizational benefit

File: scylla-server-rust/src/reciever/mock_reciever.rs (hunk 7 above)
Comment (RChandler234) [reply scylla-server-rust/src/reciever/mock_reciever.rs]: @jr1221 thoughts on this?

Comment (jr1221) [reply scylla-server-rust/src/reciever/mock_reciever.rs]: I'll try it don't see why it wouldn't work it'd just be a pub const.

What file name/location makes most sense?

Comment (RChandler234) [reply scylla-server-rust/src/reciever/mock_reciever.rs]: just this folder, and mock_data.rs

File: docker-compose-dev.yml (hunk 2 above)
Comment (jr1221) [reply docker-compose-dev.yml:76]: Tried some new docker stuff with profiles to hopefully clean up the docker compose mess lol.

File: scylla-server-rust/src/processors/mqtt_processor.rs (hunk 18 above)
Comment (jr1221) [reply scylla-server-rust/src/processors/mqtt_processor.rs]: This is if the time is literally an impossible timestamp.  Like corrupted or something.  I feel like then we shouldnt even process the message.

File: docker-compose.yml (hunk 10 above)
Comment (jr1221) [reply docker-compose.yml]: Couldnt find a way around this.  Ultimately the problem lies in argos being a hybrid package of two seperate services, scylla and the client.  ghcr doesnt want two projects to be in the same repository.

Comment (jr1221) [reply docker-compose.yml]: ya seems impossible.

File: scylla-server-rust/src/processors/mqtt_processor.rs (hunk 18 above)
Comment (Peyton-McKee) [reply scylla-server-rust/src/processors/mqtt_processor.rs]: Eh i disagree we should t try to be as resilient as possible to not lose any data. Theres checksums built into mqtt i think

File: scylla-server-rust/src/processors/mqtt_processor.rs (hunk 19 above)
Comment (Peyton-McKee) [reply scylla-server-rust/src/processors/mqtt_processor.rs]: Then L ig i mean add another check for it

**Review by Peyton-McKee — COMMENTED** [review]
Tested on router today. Slack has videos of performance. appears to be doing well with hv off. more testing required with hv on and car on jacks

**Comment by jr1221** [comment]
> Tested on router today. Slack has videos of performance. appears to be doing well with hv off. more testing required with hv on and car on jacks

Would HV on be a big difference?

**Pull request merged**